File: swiftnav_console/advanced_spectrum_analyzer_tab.py
# Copyright (c) 2022 Swift Navigation
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
# the Software, and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

"""Solution Velocity Tab QObjects.
"""

from typing import Dict, List, Any

from PySide6 import QtCharts
from PySide6.QtCore import Property, QObject, QPointF, Signal, Slot

from .constants import Keys, QTKeys


def advanced_spectrum_analyzer_tab_update() -> Dict[str, Any]:
    return {
        Keys.CHANNEL: 0,
        Keys.POINTS: [],
        Keys.YMAX: 0,
        Keys.YMIN: 0,
        Keys.XMAX: 0,
        Keys.XMIN: 0,
    }


ADVANCED_SPECTRUM_ANALYZER_TAB: List[Dict[str, Any]] = [advanced_spectrum_analyzer_tab_update()]


class AdvancedSpectrumAnalyzerPoints(QObject):
    _instance: "AdvancedSpectrumAnalyzerPoints"
    _points: List[QPointF] = []
    _ymin: float = 0.0
    _ymax: float = 0.0
    _xmin: float = 0.0
    _xmax: float = 0.0
    _channel: int = 0
    _data_updated = Signal()
    advanced_spectrum_analyzer_tab: Dict[str, Any] = {}

    def __init__(self):
        super().__init__()
        assert getattr(self.__class__, "_instance", None) is None
        self.__class__._instance = self
        self.advanced_spectrum_analyzer_tab = ADVANCED_SPECTRUM_ANALYZER_TAB[0]
        self._data_updated.connect(self.handle_data_updated)

    @classmethod
    def post_data_update(cls, update_data: Dict[str, Any]) -> None:
        ADVANCED_SPECTRUM_ANALYZER_TAB[0] = update_data
        cls._instance._data_updated.emit()  # pylint: disable=protected-access

    @Slot()  # type: ignore
    def handle_data_updated(self) -> None:
        self.advanced_spectrum_analyzer_tab = ADVANCED_SPECTRUM_ANALYZER_TAB[0]
        self.update()  # type: ignore

    def get_ymin(self) -> float:
        """Getter for _ymin."""
        return self._ymin

    def set_ymin(self, ymin: float) -> None:
        """Setter for _ymin."""
        self._ymin = ymin

    ymin = Property(float, get_ymin, set_ymin)

    def get_ymax(self) -> float:
        """Getter for _ymax."""
        return self._ymax

    def set_ymax(self, ymax: float) -> None:
        """Setter for _ymax."""
        self._ymax = ymax

    ymax = Property(float, get_ymax, set_ymax)

    def get_xmin(self) -> float:
        """Getter for _xmin."""
        return self._xmin

    def set_xmin(self, xmin: float) -> None:
        """Setter for _xmin."""
        self._xmin = xmin

    xmin = Property(float, get_xmin, set_xmin)

    def get_xmax(self) -> float:
        """Getter for _xmax."""
        return self._xmax

    def set_xmax(self, xmax: float) -> None:
        """Setter for _xmax."""
        self._xmax = xmax

    xmax = Property(float, get_xmax, set_xmax)

    def get_channel(self) -> int:
        """Getter for _channel."""
        return self._channel

    def set_channel(self, channel: int) -> None:
        """Setter for _channel."""
        self._channel = channel

    channel = Property(int, get_channel, set_channel)

    def get_points(self) -> List[QPointF]:
        return self._points

    def set_points(self, points) -> None:
        self._points = [QPointF(point.x, point.y) for point in points]

    points = Property(QTKeys.QVARIANTLIST, get_points, set_points)  # type: ignore

    @Slot(QtCharts.QXYSeries)  # type: ignore
    def fill_series(self, series):
        series.replace(self._points)


class AdvancedSpectrumAnalyzerModel(QObject):  # pylint: disable=too-few-public-methods
    @Slot(AdvancedSpectrumAnalyzerPoints)  # type: ignore
    def fill_console_points(self, cp: AdvancedSpectrumAnalyzerPoints) -> AdvancedSpectrumAnalyzerPoints:
        cp.set_points(cp.advanced_spectrum_analyzer_tab[Keys.POINTS])
        cp.set_ymax(cp.advanced_spectrum_analyzer_tab[Keys.YMAX])
        cp.set_ymin(cp.advanced_spectrum_analyzer_tab[Keys.YMIN])
        cp.set_xmax(cp.advanced_spectrum_analyzer_tab[Keys.XMAX])
        cp.set_xmin(cp.advanced_spectrum_analyzer_tab[Keys.XMIN])
        cp.set_channel(cp.advanced_spectrum_analyzer_tab[Keys.CHANNEL])
        return cp

File: console_backend/src/tabs/advanced_tab/advanced_spectrum_analyzer_tab.rs
// Copyright (c) 2022 Swift Navigation
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

use capnp::message::Builder;
use log::error;

use crate::client_sender::BoxedClientSender;
use crate::constants::{AMPLITUDES, CHANNELS, FREQUENCIES};
use crate::fft_monitor::FftMonitor;
use crate::shared_state::{SharedState, TabName};
use crate::types::Specan;
use crate::utils::serialize_capnproto_builder;

pub struct AdvancedSpectrumAnalyzerTab {
    /// Client Sender channel for communication from backend to frontend.
    client_sender: BoxedClientSender,
    /// Instance of the FftMonitor struct for handling specan messages.
    fft_monitor: FftMonitor,
    /// Stores the index received from the frontend for which CHANNEL to send to frontend.
    channel_idx: usize,
    /// Stores the currently viewed channel's amplitude values to be sent to frontend.
    most_recent_amplitudes: Vec<f32>,
    /// Stores the currently viewed channel's frequency values to be sent to frontend.
    most_recent_frequencies: Vec<f32>,
    /// The shared state for communicating between frontend/backend/other backend tabs.
    shared_state: SharedState,
}

impl AdvancedSpectrumAnalyzerTab {
    pub fn new(
        shared_state: SharedState,
        client_sender: BoxedClientSender,
    ) -> AdvancedSpectrumAnalyzerTab {
        let mut fft_monitor = FftMonitor::new();
        fft_monitor.enable_channel(None);
        AdvancedSpectrumAnalyzerTab {
            fft_monitor,
            client_sender,
            channel_idx: 0,
            most_recent_amplitudes: vec![],
            most_recent_frequencies: vec![],
            shared_state,
        }
    }

    /// Handler for Specan messages.
    ///
    /// # Parameters
    /// - `msg`: MsgSpecan to extract data from.
    pub fn handle_specan(&mut self, msg: Specan) {
        if let Err(err) = self.fft_monitor.capture_fft(msg) {
            error!("{err}");
            return;
        }
        {
            let shared_data = self.shared_state.lock();
            self.channel_idx = shared_data.advanced_spectrum_analyzer_tab.channel_idx as usize;
        }
        let channel = CHANNELS[self.channel_idx];
        if let Some(num_ffts) = self.fft_monitor.num_ffts(channel) {
            if num_ffts > 0 {
                if let Some(mut ffts) = self.fft_monitor.get_ffts(channel) {
                    if let Some(most_recent_fft) = ffts.pop() {
                        self.fft_monitor.clear_ffts(None);
                        if let Some(amplitudes) = most_recent_fft.get(&String::from(AMPLITUDES)) {
                            self.most_recent_amplitudes.clone_from(amplitudes);
                        }
                        if let Some(frequencies) = most_recent_fft.get(&String::from(FREQUENCIES)) {
                            self.most_recent_frequencies.clone_from(frequencies);
                        }

                        self.send_data();
                    }
                }
            }
        }
    }

    // /// Package data into a message buffer and send to frontend.
    pub fn send_data(&mut self) {
        if self.shared_state.current_tab() != TabName::Advanced {
            return;
        }
        let mut builder = Builder::new_default();
        let msg = builder.init_root::<crate::console_backend_capnp::message::Builder>();

        let mut tab_status = msg.init_advanced_spectrum_analyzer_status();

        let mut xmin = f32::MAX;
        let mut xmax = f32::MIN;
        let mut ymin = f32::MAX;
        let mut ymax = f32::MIN;

        let mut point_vals = tab_status
            .reborrow()
            .init_data(self.most_recent_amplitudes.len() as u32);
        for idx in 0..self.most_recent_amplitudes.len() {
            let mut point_val = point_vals.reborrow().get(idx as u32);
            point_val.set_x(self.most_recent_frequencies[idx] as f64);
            point_val.set_y(self.most_recent_amplitudes[idx] as f64);
            ymin = f32::min(self.most_recent_amplitudes[idx], ymin);
            ymax = f32::max(self.most_recent_amplitudes[idx], ymax);
            xmin = f32::min(self.most_recent_frequencies[idx], xmin);
            xmax = f32::max(self.most_recent_frequencies[idx], xmax);
        }

        tab_status.set_channel(self.channel_idx as u16);
        tab_status.set_ymin(ymin);
        tab_status.set_ymax(ymax);
        tab_status.set_xmin(xmin);
        tab_status.set_xmax(xmax);

        self.client_sender
            .send_data(serialize_capnproto_builder(builder));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{client_sender::TestSender, constants::SIGNALS_TOTAL};
    use sbp::messages::{gnss::GpsTime, piksi::MsgSpecan};

    #[test]
    fn handle_specan_empty_amplitude_value_test() {
        let shared_state = SharedState::new();
        let client_send = TestSender::boxed();
        let mut tab = AdvancedSpectrumAnalyzerTab::new(shared_state, client_send);
        let channel_tag = 2;
        let wn = 1000;
        let tow = 10001;
        let ns_residual = 100011;
        let t = GpsTime {
            tow,
            ns_residual,
            wn,
        };
        let freq_ref = 1.0;
        let freq_step = 2.0;
        let amplitude_ref = 3.0;
        let amplitude_unit = 4.0;

        let msg = MsgSpecan {
            sender_id: Some(1337),
            channel_tag,
            t,
            freq_ref,
            freq_step,
            amplitude_ref,
            amplitude_unit,
            amplitude_value: vec![],
        };
        assert!(tab.most_recent_amplitudes.is_empty());
        assert!(tab.most_recent_frequencies.is_empty());
        tab.handle_specan(Specan::MsgSpecan(msg));
        // Test against not enough amplitude value samples to trigger an update of current amps/freqs.
        assert!(tab.most_recent_amplitudes.is_empty());
        assert!(tab.most_recent_frequencies.is_empty());
    }
    #[test]
    fn handle_specan_test() {
        let shared_state = SharedState::new();
        let client_send = TestSender::boxed();
        let mut tab = AdvancedSpectrumAnalyzerTab::new(shared_state, client_send);
        let channel_tag = 1;
        let wn = 1000;
        let tow = 10001;
        let ns_residual = 100011;
        let t = GpsTime {
            tow,
            ns_residual,
            wn,
        };
        let freq_ref = 1.0;
        let freq_step = 2.0;
        let amplitude_ref = 3.0;
        let amplitude_unit = 4.0;

        let msg = MsgSpecan {
            sender_id: Some(1337),
            channel_tag,
            t,
            freq_ref,
            freq_step,
            amplitude_ref,
            amplitude_unit,
            amplitude_value: (0..SIGNALS_TOTAL).map(|i| i as u8).collect(),
        };
        assert!(tab.most_recent_amplitudes.is_empty());
        assert!(tab.most_recent_frequencies.is_empty());
        tab.handle_specan(Specan::MsgSpecan(msg));
        // Test enough samples were provided to trigger an update of current amps/freqs.
        assert_eq!(tab.most_recent_amplitudes.len(), SIGNALS_TOTAL);
        assert_eq!(tab.most_recent_frequencies.len(), SIGNALS_TOTAL);
    }
}
